executable-spec/src/Cardano/Ledger/Spec/STS/Chain/Chain.hs

```diff
@@ -0,0 +1,152 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedLists #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+
+-- | Simple block-chain mock, to bundle transactions into blocks, including slot
+-- ticks.
+module Cardano.Ledger.Spec.STS.Chain.Chain where
+
+import           Data.Function ((&))
+import           Data.Map.Strict (Map)
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Range as Range
+
+import           Cardano.Prelude (HeapWords, heapWords, heapWords2)
+
+import           Control.State.Transition (Embed, Environment, IRC (IRC),
+                     PredicateFailure, STS, Signal, State, TRC (TRC),
+                     Threshold (Threshold), initialRules, judgmentContext,
+                     trans, transitionRules, wrapFailed, (?!))
+import           Control.State.Transition.Generator (HasTrace, envGen, sigGen)
+
+import           Ledger.Core (Slot (Slot))
+
+import           Cardano.Ledger.Spec.STS.Sized (WordCount, size)
+import           Cardano.Ledger.Spec.STS.Transaction.Transaction (TRANSACTIONS)
+import qualified Cardano.Ledger.Spec.STS.Transaction.Transaction as Dummy
+
+
+data CHAIN
+
+data St
+  = St
+    { currentSlot :: !Slot
+    , submitted :: Map Dummy.Transaction Slot
+    }
+  deriving (Eq, Show)
+
+
+data Env
+  = Env
+    { initialSlot :: !Slot
+    , maximumBlockSize :: !WordCount
+    -- ^ Maximum block size. For now we measure this in number of 'Word's using
+    -- 'heapWords' from 'Cardano.Prelude.HeapWords'.
+    }
+  deriving (Eq, Show)
+
+data Block
+  = Block
+    { slot :: !Slot
+    , transactions :: ![Transaction]
+    }
+    deriving (Eq, Show)
+
+instance HeapWords Block where
+  heapWords (Block (Slot s) transactions) = heapWords2 s transactions
+
+data Transaction = Dummy Dummy.Transaction
```

Comment:
Good question. I haven't consider using a type-class for the block payload, however I think:

```haskell
data Tx a => Block a = 
  Block
    { slot :: !Slot
    , transactions :: ![a]
    }
    deriving (Eq, Show)
```

Won't work since you can only have one type of payload (but I haven't tried this myself, so I could be wrong).

About the sum type, we might want to do something like:

```haskell
data Transaction = Dummy ... | Update ...
```

and have `Update` be a sum type.